Regression: NoSuchIndex is now a DatabaseError

I can't seem to repro this on console.neo4j.org, but at least locally, running Neo4j 2.2.6, when I make the following query via the `/db/data/transaction` REST API:

``` cypher
MATCH (n:Test)
USING INDEX n:Test(foo)
WHERE n.foo = {foo}
RETURN n
```

I get back a `Neo.DatabaseError.Statement.ExecutionFailure` instead of what used to be a `Neo.ClientError.Schema.NoSuchIndex`.

The error message and stack trace (pretty-printed below) look like the underlying error is still the same; it maybe just isn't being properly handled?

```
org.neo4j.kernel.impl.query.QueryExecutionKernelException: No such index found.
Label: `Test`
Property name: `index_09002856723964214`
    at org.neo4j.cypher.javacompat.internal.ServerExecutionEngine.executeQuery(ServerExecutionEngine.java:65)
    at org.neo4j.server.rest.transactional.TransactionHandle.executeStatements(TransactionHandle.java:327)
    at org.neo4j.server.rest.transactional.TransactionHandle.commit(TransactionHandle.java:146)
    at org.neo4j.server.rest.web.TransactionalService$2.write(TransactionalService.java:211)
    at com.sun.jersey.core.impl.provider.entity.StreamingOutputProvider.writeTo(StreamingOutputProvider.java:71)
    at com.sun.jersey.core.impl.provider.entity.StreamingOutputProvider.writeTo(StreamingOutputProvider.java:57)
    at com.sun.jersey.spi.container.ContainerResponse.write(ContainerResponse.java:302)
    at com.sun.jersey.server.impl.application.WebApplicationImpl._handleRequest(WebApplicationImpl.java:1510)
    at com.sun.jersey.server.impl.application.WebApplicationImpl.handleRequest(WebApplicationImpl.java:1419)
    at com.sun.jersey.server.impl.application.WebApplicationImpl.handleRequest(WebApplicationImpl.java:1409)
    at com.sun.jersey.spi.container.servlet.WebComponent.service(WebComponent.java:409)
    at com.sun.jersey.spi.container.servlet.ServletContainer.service(ServletContainer.java:558)
    at com.sun.jersey.spi.container.servlet.ServletContainer.service(ServletContainer.java:733)
    at javax.servlet.http.HttpServlet.service(HttpServlet.java:790)
    at org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:800)
    at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:587)
    at org.eclipse.jetty.server.session.SessionHandler.doHandle(SessionHandler.java:221)
    at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1125)
    at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:515)
    at org.eclipse.jetty.server.session.SessionHandler.doScope(SessionHandler.java:185)
    at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1059)
    at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:141)
    at org.eclipse.jetty.server.handler.HandlerList.handle(HandlerList.java:52)
    at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:97)
    at org.eclipse.jetty.server.Server.handle(Server.java:497)
    at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:310)
    at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:248)
    at org.eclipse.jetty.io.AbstractConnection$2.run(AbstractConnection.java:540)
    at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:620)
    at org.eclipse.jetty.util.thread.QueuedThreadPool$3.run(QueuedThreadPool.java:540)
    at java.lang.Thread.run(Thread.java:745)
Caused by: org.neo4j.cypher.IndexHintException: No such index found.
Label: `Test`
Property name: `index_09002856723964214`
    at org.neo4j.cypher.internal.compatibility.exceptionHandlerFor2_2$.indexHintException(CompatibilityFor2_2.scala:102)
    at org.neo4j.cypher.internal.compatibility.exceptionHandlerFor2_2$.indexHintException(CompatibilityFor2_2.scala:53)
    at org.neo4j.cypher.internal.compiler.v2_2.IndexHintException.mapToPublic(CypherException.scala:92)
    at org.neo4j.cypher.internal.compatibility.exceptionHandlerFor2_2$.runSafely(CompatibilityFor2_2.scala:113)
    at org.neo4j.cypher.internal.compatibility.CompatibilityFor2_2$$anon$2.plan(CompatibilityFor2_2.scala:159)
    at org.neo4j.cypher.ExecutionEngine$$anonfun$liftedTree1$1$1$$anonfun$apply$1.apply(ExecutionEngine.scala:154)
    at org.neo4j.cypher.ExecutionEngine$$anonfun$liftedTree1$1$1$$anonfun$apply$1.apply(ExecutionEngine.scala:151)
    at org.neo4j.cypher.internal.compiler.v2_2.MonitoringCacheAccessor$$anonfun$1.apply(CacheAccessor.scala:35)
    at org.neo4j.cypher.internal.compiler.v2_2.helpers.LRUCache.getOrElseUpdate(LRUCache.scala:34)
    at org.neo4j.cypher.internal.compiler.v2_2.helpers.LRUCache.apply(LRUCache.scala:64)
    at org.neo4j.cypher.internal.compiler.v2_2.MonitoringCacheAccessor.getOrElseUpdate(CacheAccessor.scala:33)
    at org.neo4j.cypher.ExecutionEngine$$anonfun$liftedTree1$1$1.apply(ExecutionEngine.scala:151)
    at org.neo4j.cypher.ExecutionEngine$$anonfun$liftedTree1$1$1.apply(ExecutionEngine.scala:151)
    at scala.collection.Iterator$$anon$9.next(Iterator.scala:162)
    at scala.collection.Iterator$$anon$13.hasNext(Iterator.scala:371)
    at scala.collection.Iterator$$anon$13.next(Iterator.scala:372)
    at org.neo4j.cypher.ExecutionEngine.liftedTree1$1(ExecutionEngine.scala:163)
    at org.neo4j.cypher.ExecutionEngine.planQuery(ExecutionEngine.scala:143)
    at org.neo4j.cypher.ExecutionEngine.execute(ExecutionEngine.scala:107)
    at org.neo4j.cypher.ExecutionEngine.execute(ExecutionEngine.scala:102)
    at org.neo4j.cypher.javacompat.internal.ServerExecutionEngine.executeQuery(ServerExecutionEngine.java:61)
    ... 30 more
```

Not sure exactly when this regression happened. Also haven't tested with 2.2.7 or 2.3.x yet.

--- a/community/cypher/cypher/src/main/scala/org/neo4j/cypher/CypherException.scala
+++ b/community/cypher/cypher/src/main/scala/org/neo4j/cypher/CypherException.scala
@@ -98,12 +98,12 @@ class UnknownLabelException(labelName: String) extends CypherException(s"The pro
 
 class HintException( message: String)
   extends CypherException(message) {
-  val status = Status.Schema.NoSuchIndex
+  val status = Status.Statement.ExecutionFailure
 }
 
 class IndexHintException(identifier: String, label: String, property: String, message: String)
   extends CypherException(s"$message\nLabel: `$label`\nProperty name: `$property`") {
-  val status = Status.Statement.ExecutionFailure
+  val status = Status.Schema.NoSuchIndex
 }
 
 class LabelScanHintException(identifier: String, label: String, message: String)

--- a/community/cypher/cypher/src/test/scala/org/neo4j/cypher/UsingAcceptanceTest.scala
+++ b/community/cypher/cypher/src/test/scala/org/neo4j/cypher/UsingAcceptanceTest.scala
@@ -19,6 +19,7 @@
  */
 package org.neo4j.cypher
 
+import org.neo4j.kernel.api.exceptions.Status
 
 class UsingAcceptanceTest extends ExecutionEngineFunSuite with NewPlannerTestSupport {
 
@@ -219,6 +220,41 @@ class UsingAcceptanceTest extends ExecutionEngineFunSuite with NewPlannerTestSup
         |RETURN 13 as id""".stripMargin)
 
     result.toList should be(empty)
+  }
+
+  test("when failing to support all hints we should provide an understandable error message") {
+    // GIVEN
+    graph.createIndex("LocTag", "id")
+
+    // WHEN
+    val query = """MATCH (t1:LocTag {id:1642})-[:Child*0..]->(:LocTag)
+                  |     <-[:Tagged]-(s1:Startup)<-[r1:Role]-(u:User)
+                  |     -[r2:Role]->(s2:Startup)-[:Tagged]->(:LocTag)
+                  |     <-[:Child*0..]-(t2:LocTag {id:1642})
+                  |USING INDEX t1:LocTag(id)
+                  |USING INDEX t2:LocTag(id)
+                  |RETURN count(u)""".stripMargin
+
+
+    val error = intercept[HintException](executeWithAllPlanners(query))
+
+    error.getMessage should equal("The current planner cannot satisfy all hints in the query, please try removing hints or try with another planner")
+    error.status should equal(Status.Statement.ExecutionFailure)
+  }
+
+  test("correct status code when no index") {
+
+    // GIVEN
+    val query = """MATCH (n:Test)
+                  |USING INDEX n:Test(foo)
+                  |WHERE n.foo = {foo}
+                  |RETURN n""".stripMargin
+
+    // WHEN
+    val error = intercept[IndexHintException](executeWithAllPlanners(query))
+
+    // THEN
+    error.status should equal(Status.Schema.NoSuchIndex)
   }
 
 }

--- a/community/server/src/test/java/org/neo4j/server/rest/transactional/integration/TransactionIT.java
+++ b/community/server/src/test/java/org/neo4j/server/rest/transactional/integration/TransactionIT.java
@@ -19,6 +19,9 @@
  */
 package org.neo4j.server.rest.transactional.integration;
 
+import org.codehaus.jackson.JsonNode;
+import org.junit.Test;
+
 import java.io.InputStream;
 import java.io.InputStreamReader;
 import java.io.PrintStream;
@@ -30,9 +33,6 @@ import java.util.concurrent.Callable;
 import java.util.concurrent.CountDownLatch;
 import java.util.concurrent.Executors;
 import java.util.concurrent.Future;
-
-import org.codehaus.jackson.JsonNode;
-import org.junit.Test;
 
 import org.neo4j.graphdb.DynamicLabel;
 import org.neo4j.graphdb.Label;
@@ -54,7 +54,6 @@ import static org.hamcrest.CoreMatchers.is;
 import static org.hamcrest.MatcherAssert.assertThat;
 import static org.junit.Assert.assertEquals;
 import static org.junit.Assert.assertTrue;
-
 import static org.neo4j.helpers.collection.IteratorUtil.asSet;
 import static org.neo4j.server.rest.domain.JsonHelper.jsonNode;
 import static org.neo4j.server.rest.transactional.integration.TransactionMatchers.containsNoErrors;
@@ -823,7 +822,7 @@ public class TransactionIT extends AbstractRestFunctionalTestBase
         // given
         final String hostname = "dummy.example.org";
         final String scheme = "http";
-        
+
         // when
         Response rs = http.withHeaders( XForwardUtil.X_FORWARD_HOST_HEADER_KEY, hostname )
                 .POST( "/db/data/transaction/commit", quotedJson(
@@ -851,6 +850,15 @@ public class TransactionIT extends AbstractRestFunctionalTestBase
         assertPath( restNode.get( "all_relationships" ), "/node/\\d+/relationships/all", hostname, scheme );
         assertPath( restNode.get( "incoming_typed_relationships" ),
                 "/node/\\d+/relationships/in/\\{-list\\|&\\|types\\}", hostname, scheme );
+    }
+
+    @Test
+    public void correctStatusCodeWhenUsingHintWithoutAnyIndex() throws Exception
+    {
+        // begin and execute and commit
+        Response begin = http.POST( "/db/data/transaction/commit", quotedJson( "{ 'statements': [ { 'statement': " +
+                                                                               "'MATCH (n:Test) USING INDEX n:Test(foo) WHERE n.foo = 42 RETURN n.foo' } ] }" ) );
+        assertThat( begin, hasErrors( Status.Request.Schema.NoSuchIndex ) );
     }
 
     private void assertPath( JsonNode jsonURIString, String path, String hostname, final String scheme )

